src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "constant_folding.hpp"
+#include "ngraph/op/non_zero.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace
+{
+    struct NonZeroElements
+    {
+        using Results_t = std::vector<std::vector<int64_t>>;
+
+        NonZeroElements(const Shape& input_shape)
+            : m_input_shape{input_shape}
+            , m_results{Results_t(input_shape.size())}
+        {
+        }
+
+        template <typename T>
+        const Results_t& find_indices(const T* values)
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
+            const auto values_count = shape_size(m_input_shape);
+
+            const T zero_value = T{0};
+            for (size_t i = 0; i < values_count; ++i)
+            {
+                if (values[i] != zero_value)
+                {
+                    add_to_results(m_current_index);
+                }
+
+                // don't generate the next index when the last value has been processed
+                if (i < values_count - 1)
```

Comment:
It will be more efficient if we change the "for" loop condition to "i + 1 < values_count" and remove the "if" statement in this line. Then after the "for" loop we check that "values_count != 0" and add if the last value is not equal to 0 then call "add_to_results(m_current_index)"

Reply:
Done, thanks for the tip